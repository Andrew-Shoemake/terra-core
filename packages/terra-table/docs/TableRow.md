# Terra Table Row

The Terra Table Row is a sub component that groups child cells. Each row handles it's own selection state through props. The onSelect function that can be passed to the table row will return the metaData associated to the row. The metaData allows for greater control of state with the ability associate any indicator or function with the row.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-table`
  - `yarn add terra-table`

## Usage

```jsx
import React from 'react';
import { Row, Cell } from 'terra-table';

  <Row
    key="unique-react-key"
    isSelectable
    isSelected
    metaData={{ myInfo: 'my-data' }}
    onSelect={this.handleRowSelection}
  >
    <Cell>
      Cell Content
    </Cell>
    <Cell>
      Cell Content
    </Cell>
  </Row>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
