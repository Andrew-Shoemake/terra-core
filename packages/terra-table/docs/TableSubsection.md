# Terra Table Subsection

The Terra Table Subsection is a sub component that groups child rows beneath a subsection header. The subsection header has the ability to toggle the display of its child rows. Each subsection handles its collapsible and collapsed state.

## Getting Started

- Install with [npmjs](https://www.npmjs.com):
  - `npm install terra-table`
  - `yarn add terra-table`

## Usage

```jsx
import React from 'react';
import { Row, Subsection } from 'terra-table';

  <Subsection
    key="unique-react-key"
    isCollapsible
    isCollapsed
    metaData={{ myInfo: 'my-data' }}
    onSelect={this.handleRowSelection}
  >
    <Row />
    <Row />
    <Row />
  </Subsection>
```

## Component Features
* [Cross-Browser Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#cross-browser-support)
* [Responsive Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#responsive-support)
* [Mobile Support](https://github.com/cerner/terra-ui/blob/master/src/terra-dev-site/contributing/ComponentStandards.e.contributing.md#mobile-support)
