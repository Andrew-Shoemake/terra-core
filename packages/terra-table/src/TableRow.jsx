import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import 'terra-base/lib/baseStyles';
import styles from './TableRow.module.scss';
import TableUtils from './TableUtils';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * The children passed to the component
   */
  children: PropTypes.node,
  /**
   * Whether or not row is selected
   */
  isSelected: PropTypes.bool,
  /**
   * Whether or not row is selectable
   */
  isSelectable: PropTypes.bool,
  /**
   * The associated metaData to be provided in the onSelect callback.
   */
  // eslint-disable-next-line react/forbid-prop-types
  metaData: PropTypes.object,
  /**
   * Function callback for when the appropriate click or key action is performed.
   * Callback contains the javascript evnt and prop metadata, e.g. onSelect(event, metaData)
   */
  onSelect: PropTypes.func,
  /**
   * Function callback for the ref of the tr.
   */
  refCallback: PropTypes.func,
};

const defaultProps = {
  isSelected: false,
  isSelectable: false,
};

const TableRow = ({
  children,
  isSelected,
  isSelectable,
  metaData,
  onBlur,
  onClick,
  onKeyDown,
  onMouseDown,
  onSelect,
  refCallback,
  ...customProps
}) => {
  const rowClassNames = cx([
    { 'is-selected': isSelected && isSelectable },
    { 'is-selectable': isSelectable },
    'row',
    customProps.className,
  ]);

  const attrSpread = {};
  if (isSelectable) {
    attrSpread.onClick = TableUtils.wrappedOnClickForItem(onClick, onSelect, metaData);
    attrSpread.onKeyDown = TableUtils.wrappedOnKeyDownForItem(onKeyDown, onSelect, metaData);
    attrSpread.tabIndex = '0';
    attrSpread['aria-selected'] = isSelected;
    attrSpread['data-row-show-focus'] = 'true';
    attrSpread.onBlur = TableUtils.wrappedEventCallback(onBlur, event => event.currentTarget.setAttribute('data-row-show-focus', 'true'));
    attrSpread.onMouseDown = TableUtils.wrappedEventCallback(onMouseDown, event => event.currentTarget.setAttribute('data-row-show-focus', 'false'));
  }

  return (
    <tr {...customProps} {...attrSpread} className={rowClassNames} ref={refCallback}>
      {children}
    </tr>
  );
};

TableRow.propTypes = propTypes;
TableRow.defaultProps = defaultProps;

export default TableRow;
