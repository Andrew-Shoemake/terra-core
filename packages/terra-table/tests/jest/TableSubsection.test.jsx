import React from 'react';
import TableSubsection from '../../src/TableSubsection';
import TableRow from '../../src/TableRow';

// Snapshot test
it('should render a default table section', () => {
  const section = shallow(
    <TableSubsection colSpan={1} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section with a row', () => {
  const section = shallow(
    <TableSubsection colSpan={1}>
      <TableRow />
    </TableSubsection>,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section with isCollpased', () => {
  const section = shallow(
    <TableSubsection colSpan={1} isCollapsed />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section with isCollapsible', () => {
  const section = shallow(
    <TableSubsection colSpan={1} isCollapsible />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section with level', () => {
  const section = shallow(
    <TableSubsection colSpan={1} level={4} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section with metaData', () => {
  const section = shallow(
    <TableSubsection colSpan={1} metaData={{}} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section with onSelect', () => {
  const section = shallow(
    <TableSubsection colSpan={1} onSelect={jest.fn()} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section with refCallback', () => {
  const section = shallow(
    <TableSubsection colSpan={1} onSelect={jest.fn()} />,
  );
  expect(section).toMatchSnapshot();
});

it('should render a table section with title', () => {
  const section = shallow(
    <TableSubsection colSpan={1} title="test" />,
  );
  expect(section).toMatchSnapshot();
});
