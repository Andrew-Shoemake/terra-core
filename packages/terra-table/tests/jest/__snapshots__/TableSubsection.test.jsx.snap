// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a default table section 1`] = `
<Fragment>
  <TableSubsectionHeader
    colSpan={1}
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a table section with a row 1`] = `
<Fragment>
  <TableSubsectionHeader
    colSpan={1}
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    title="test"
  />
  <TableRow
    isSelectable={false}
    isSelected={false}
    isStriped={false}
  />
</Fragment>
`;

exports[`should render a table section with isCollapsible 1`] = `
<Fragment>
  <TableSubsectionHeader
    colSpan={1}
    isCollapsed={false}
    isCollapsible={true}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a table section with isCollpased 1`] = `
<Fragment>
  <TableSubsectionHeader
    colSpan={1}
    isCollapsed={true}
    isCollapsible={false}
    level={3}
    title="test"
  />
</Fragment>
`;

exports[`should render a table section with level 1`] = `
<Fragment>
  <TableSubsectionHeader
    colSpan={1}
    isCollapsed={false}
    isCollapsible={false}
    level={4}
    title="test"
  />
</Fragment>
`;

exports[`should render a table section with metaData 1`] = `
<Fragment>
  <TableSubsectionHeader
    colSpan={1}
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    metaData={Object {}}
    title="test"
  />
</Fragment>
`;

exports[`should render a table section with onSelect 1`] = `
<Fragment>
  <TableSubsectionHeader
    colSpan={1}
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    onSelect={[MockFunction]}
    title="test"
  />
</Fragment>
`;

exports[`should render a table section with refCallback 1`] = `
<Fragment>
  <TableSubsectionHeader
    colSpan={1}
    isCollapsed={false}
    isCollapsible={false}
    level={3}
    refCallback={[MockFunction]}
    title="test"
  />
</Fragment>
`;
